packages/controls/src/controls/shape-v2/shape-v2.ts

```diff
@@ -0,0 +1,322 @@
+          layout: z
+            .union([
+              z.literal(Definition.Layout.Inline),
+              z.literal(Definition.Layout.Popover),
+            ])
+            .optional(),
+          type: z.record(z.string(), SerializationSchema.deserializedRecord),
+        }),
+      })
+      .parse(data)
+
+    const deserializedType = mapValues(type, (itemDef) => {
+      return deserializeCallback(itemDef)
+    })
+
+    return ShapeV2({
+      type: deserializedType,
+      ...config,
+    })
+  }
+
+  get keyDefs() {
+    return this.config.type
+  }
+
+  get controlType() {
+    return Definition.type
+  }
+
+  get schema() {
+    const version = z.literal(1)
+    const type = z.literal(Definition.type)
+
+    const keys = mapValues(this.keyDefs, (def) => def.schema)
+
+    const v0Data = z.object(mapValues(keys, (key) => key.data)) as SchemaType<
+      ShapeV2DataTypeV0<C>
+    >
+    const v1Data = z.object({
+      [ControlDataTypeKey]: z.literal(Definition.v1DataType),
+      value: v0Data,
+    }) as SchemaType<ShapeV2DataTypeV1<C>>
+
+    const data = z.union([v0Data, v1Data])
+
+    const value = z.object(mapValues(keys, (key) => key.value)) as SchemaType<
+      ValueType<C>
+    >
+
+    const resolvedValue = z.object(
+      mapValues(keys, (key) => key.resolvedValue),
+    ) as SchemaType<ResolvedValueType<C>>
+
+    const config = z.object({
+      label: z.string().optional(),
+      layout: z.union([
+        z.literal(Definition.Layout.Inline),
+        z.literal(Definition.Layout.Popover),
+      ]),
+      type: z.object(mapValues(keys, (key) => key.definition)),
+    })
+
+    const definition = z.object({
+      type,
+      config,
+    })
+
+    return { type, data, value, resolvedValue, definition, version }
+  }
+
+  safeParse(data: unknown | undefined): ParseResult<DataType<C> | undefined> {
+    return safeParse(this.schema.data, data)
+  }
+
+  fromData(data: DataType<C> | undefined): ValueType<C> | undefined {
+    if (data == null) return undefined
+
+    if (
+      ControlDataTypeKey in data &&
+      data[ControlDataTypeKey] === Definition.v1DataType
+    ) {
+      return mapValues(this.keyDefs, (def, key) =>
+        def.fromData((data as ShapeV2DataTypeV1<C>).value[key]),
+      )
+    }
+
+    return mapValues(this.keyDefs, (def, key) =>
+      def.fromData((data as ShapeV2DataTypeV0<C>)[key]),
+    )
+  }
+
+  toData(value: ValueType<C>): DataType<C> {
+    return {
+      ...Definition.dataSignature.v1,
+      value: mapValues(this.keyDefs, (def, key) => def.toData(value[key])),
+    }
+  }
+
+  copyData(
+    data: DataType<C> | undefined,
+    context: CopyContext,
+  ): DataType<C> | undefined {
+    if (data == null) return undefined
+
+    if (
+      ControlDataTypeKey in data &&
+      data[ControlDataTypeKey] === Definition.v1DataType
+    ) {
+      return mapValues(this.keyDefs, (def, key) =>
+        def.copyData((data as ShapeV2DataTypeV1<C>).value[key], context),
+      )
```

Comment:
Same here: https://github.com/makeswift/makeswift/blob/466ed3b357f277fe17185822ce9320d9153c84ea/packages/controls/src/controls/image/image.ts#L265

Reply:
As long as we agree on the `toShapeData` I have above then all of these such locations are now using that and don't have branching.